sklearn/feature_selection/tests/test_mutual_info.py

```diff
@@ -0,0 +1,171 @@
+from __future__ import division
+
+import numpy as np
+from numpy.testing import run_module_suite
+from scipy.sparse import csr_matrix
+
+from sklearn.utils.testing import (assert_array_equal, assert_almost_equal,
+                                   assert_false, assert_true, assert_raises)
+from sklearn.feature_selection.mutual_info_ import (
+    mutual_info_regression, mutual_info_classif, _compute_mi)
+
+
+class TestMIComputation(object):
+    def test_dd(self):
+        # In discrete case computations are straightforward and can be done
+        # by hand on given vectors.
+        x = np.array([0, 1, 1, 0, 0])
+        y = np.array([1, 0, 0, 0, 1])
+
+        H_x = H_y = -(3/5) * np.log(3/5) - (2/5) * np.log(2/5)
+        H_xy = -1/5 * np.log(1/5) - 2/5 * np.log(2/5) - 2/5 * np.log(2/5)
+        I_xy = H_x + H_y - H_xy
+
+        assert_almost_equal(_compute_mi(x,  y, True, True), I_xy)
+
+    def test_cc(self):
+        # For two continuous variables a good approach is to test on bivariate
+        # normal distribution, where mutual information is known.
+
+        # Mean of the distribution, irrelevant for mutual information.
+        mean = np.zeros(2)
+
+        # Setup covariance matrix with correlation coeff. equal 0.5.
+        sigma_1 = 1
+        sigma_2 = 10
+        corr = 0.5
+        cov = np.array([
+            [sigma_1**2, corr * sigma_1 * sigma_2],
+            [corr * sigma_1 * sigma_2, sigma_2**2]
+        ])
+
+        # True theoretical mutual information.
+        I_theory = (np.log(sigma_1) + np.log(sigma_2) -
+                    0.5 * np.log(np.linalg.det(cov)))
+
+        np.random.seed(0)
+        Z = np.random.multivariate_normal(mean, cov, size=1000)
+
+        x, y = Z[:, 0], Z[:, 1]
+
+        # Theory and computed values won't be very close, assert that relative
+        # error is less than 10%.
+        for n_neighbors in [3, 5, 7]:
+            I_computed = _compute_mi(x, y, False, False, n_neighbors)
+            assert_true(np.abs(I_computed - I_theory) < 0.1 * I_theory)
+
+    def test_cd(self):
+        # To test define a joint distribution as follows:
+        # p(x, y) = p(x) p(y | x)
+        # X ~ Bernoulli(p)
+        # (Y | x = 0) ~ Uniform(-1, 1)
+        # (Y | x = 1) ~ Uniform(0, 2)
+
+        # Use the following formula for mutual information:
+        # I(X; Y) = H(Y) - H(Y | X)
+        # Two entropies can be computed by hand:
+        # H(Y) = -(1-p)/2 * ln((1-p)/2) - p/2*log(p/2) - 1/2*log(1/2)
+        # H(Y | X) = ln(2)
```

Comment:
Which means when

1. -1 < y < 0, H(Y|X) = P(X = 0) H(Y | X = 0) = (1 - p) * 0.5 * ln(2)
2.  0 < y < 1 is as before
3.  1 < y < 2. H(Y|X) = P(X = 1) H(Y | X = 1) = p * 0.5 * ln(2)

and all other values of y, it would be zero